Automatically load PuTTY SSH key files when needed in left panel

mast-eu opened the pull request:
This is a follow-up of #6293 and needs to be rebased before it can be merged.


## Proposed changes

- copy `LoadPuttyKey()` from `FormPull.cs` into `RepoObjectsTree.cs` in order to automatically load SSH keys (if available) when `Update all remotes`, `Update and prune all remotes` or `Prune remote` is clicked in the left panel (#6293 adds these to the left panel context menu)


## Test methodology

- manual tests
- ran unit tests


## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 3.00.00.4433
- Git Extensions 3.1.0
- Build 5cc1e8e40e59ce000aba01ce5351ebeb01343144
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.3324.0
- DPI 120dpi (125% scaling)



----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

RussKie commented:
Could you please rebase and fix the MC

mast-eu commented:
> Could you please rebase and fix the MC

done

RussKie commented on GitUI/BranchTreePanel/RepoObjectsTree.cs:
```diff
@@ -472,17 +474,72 @@ public TestAccessor(RepoObjectsTree repoObjectsTree)
 
         private void mnuBtnPrune_Click(object sender, EventArgs e)
         {
+            LoadPuttyKey();
```
`LoadPuttyKey` is run before each `FormRemoteProcess.ShowDialog` thus indicating it is a concern better placed with `FormRemoteProcess`.
It also doesn't seem correct to make the left panel responsible for loading or otherwise managing SSH keys.

`FormRemoteProcess` already has some functionality that looks like it interacts with Plink, e.g.
https://github.com/gitextensions/gitextensions/blob/027eabec3be497f8d780cc68ece268c64a43a7d5/GitUI/FormRemoteProcess.cs#L145-L153

pmiossec replied:
I also think that we should not duplicate `LoadPuttyKey()` code and do a refactoring to prevent it....

mast-eu replied:
Actually, there is already a good working implementation in place for "Fetch" (i.e. the option that was already implemented before #6293):

![](https://user-images.githubusercontent.com/460196/53209112-afae9200-3639-11e9-9ae4-04c8bdbd658d.png)

"Fetch" is connected to:
https://github.com/gitextensions/gitextensions/blob/96090c8d0114397fb0bcfbbb342261d1da45e904/GitUI/BranchTreePanel/RepoObjectsTree.Nodes.Remotes.cs#L349-L357

Which uses the pull dialog, including it's `LoadPuttyKey()`:
https://github.com/gitextensions/gitextensions/blob/96090c8d0114397fb0bcfbbb342261d1da45e904/GitUI/CommandsDialogs/FormPull.cs#L739

I propose to extend the same mechanism for all newly added options. I have to modify  @pmiossec's implementation, but I think it's worth it. No code duplication and all options to fetch/update/prune remotes from the left panel will work the same way.

mast-eu commented:
## Proposed changes

- Left panel remotes context menu: changed "Update all remotes" to "Fetch all remotes" (both text and git command); added icon (same as corresponding pull dialog shortcut)
- Left panel remotes context menu: changed "Update and prune all remotes" to "Fetch and prune all remotes" (both text and git command); added icon (same as corresponding pull dialog shortcut)
- Left panel remote context menu: changed icon for "Fetch" (same as corresponding pull dialog shortcut)
- Left panel remote context menu : changed "Prune remote" to "Fetch and prune" (both text and git command); created new icon, similar to the other ones
- all above actions now use the existing pull dialog, thus avoid code duplication and keep logic for handling SSH keys away from the left panel
- refactoring of the MessageBox "Warning! The fetch with prune will remove all the remote-tracking references which no longer exist on remotes. Do you want to proceed?", in order to show it also when pruning is triggered from the left panel (via "Fetch and prune all remotes" or "Fetch and prune")


## Screenshots

### Before
![image](https://user-images.githubusercontent.com/460196/53209191-ff8d5900-3639-11e9-904d-3aeeca559b5e.png)

 ![image](https://user-images.githubusercontent.com/460196/53209123-bccb8100-3639-11e9-9a21-ec3ae26cab50.png)

![oldmessage](https://user-images.githubusercontent.com/46861028/53671977-f37c4980-3c80-11e9-8ee4-7979bcf45322.png)

### After
![contextall](https://user-images.githubusercontent.com/46861028/53672090-4fdf6900-3c81-11e9-9df9-522098468e4a.png)

![contextsingle](https://user-images.githubusercontent.com/46861028/53672089-4fdf6900-3c81-11e9-8a90-2ca73a4ed619.png)

![newmessageall](https://user-images.githubusercontent.com/46861028/53671997-10b11800-3c81-11e9-9f44-81336cc3d23c.png)

![newmessagesingle](https://user-images.githubusercontent.com/46861028/53671998-10b11800-3c81-11e9-8a5d-65ba341afb73.png)


## Test methodology

- manual tests
- ran unit tests


## Test environment(s) <!-- Remove any that don't apply -->

- Git Extensions 3.1.0
- Build 34bc9550a47a3c50aacd4b5faf6352719240efcd
- Git 2.20.1.windows.1
- Microsoft Windows NT 6.1.7601 Service Pack 1
- .NET Framework 4.7.3324.0
- DPI 120dpi (125% scaling)


----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

mast-eu commented:
@pmiossec  I changed quite a bit what you have implemented in #6293. Could you please have a look?

RussKie commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -691,6 +691,7 @@ public enum PullAction
             Merge,
             Rebase,
             Fetch,
+            FetchPrune,
```
❗️ This will have a profound effect on the default pull action dropdown. 
We can't do that.

![image](https://user-images.githubusercontent.com/4403806/53674952-e6457600-3ce5-11e9-849d-fd321c4b3062.png)

mast-eu replied:
I didn't find any problem (yet). What I tested so far:
1) set all available default pull actions and verify that they are saved correctly in GitExtensions.settings item `DefaultPullAction` -> ok
2) set all available default pull actions and verify that they are applied correctly after clicking `Open pull dialog ...` and the dialog is setup correctly -> ok
3) trigger all available options from the pull dialog dropdown and verify that they execute the expected git command -> ok
4) trigger all available options from the left panel and verify that they execute the expected git command -> ok
5) Extended unit tests `Should_correctly_setup_for_defined_pull_action` and `Should_use_user_DefaultPullAction_pull_action_None` wich cases for `PullAction.FetchPrune` -> ok

Do I miss something?

Would it be better to add the new "Fetch and prune" to the default pull action dropdown as well?

RussKie replied:
These options are also used when the app is invoked via the shell extensions.
We had a number of issues associated with this area, e.g. #6144, #6194, #6200; and I wouldn't want a repeat of these...

mast-eu replied:
I force pushed changes to remove `PullAction.FetchPrune`.

RussKie commented on GitUI/CommandsDialogs/FormPull.cs:
```diff
@@ -97,6 +97,8 @@ public sealed partial class FormPull : GitModuleForm
         private readonly TranslationString _buttonPull = new TranslationString("&Pull");
         private readonly TranslationString _buttonFetch = new TranslationString("&Fetch");
+
+        private readonly TranslationString _pullFetchPruneAllConfirmation = new TranslationString("Warning! The fetch with prune will remove all the remote-tracking references which no longer exist on remotes. Do you want to proceed?");
```
The display of the warning must be configurable
https://github.com/gitextensions/gitextensions/blob/f6929f477c8194dcf3ad3179df3572b9029bf18a/GitUI/CommandsDialogs/FormBrowse.cs#L2421-L2448

mast-eu replied:
Display of the warning is controlled by `AppSettings.DontConfirmFetchAndPruneAll`, which can be set in `Settings` > `Advanced` > `Confirmations`. I didn't change anything in this regard.

![fetchpruneall](https://user-images.githubusercontent.com/46861028/53687896-208d3280-3d3b-11e9-8ec3-2a5d05a30e15.png)

Thinking about it, maybe the text should be changed to something like " Fetch and prune branches", because now it applies also to a single repo and no longer necessarily to all repos at once?

RussKie replied:
Confirm that it is working as intended, thank you

RussKie replied:
> maybe the text should be changed to something like "Fetch and prune branches",

Sounds good, the suggestion is better than the existing text.

RussKie commented on GitUI/CommandsDialogs/FormPull.cs:
```diff
@@ -216,8 +222,35 @@ private void BindRemotesDropDown(string selectedRemoteName)
+
+            if (pullAction == AppSettings.PullAction.FetchPruneAll)
+            {
+                messageBoxTitle = string.Format(_pruneFromCaption.Text, AllRemotes);
```
Can we express the difference under `PullAction.FetchPruneAll` but supply `AllRemotes` if `remote == null`?

mast-eu replied:
Sorry, I'm not sure if I understand you correctly. Do you want to simply improve the readability of this method?

Or do you want to eliminate `PullAction.FetchPrune` by using `PullAction.FetchPruneAll` for both operations "Fetch and prune" and "Fetch and prune all"? Something like:
```
    if (pullAction == AppSettings.PullAction.FetchPruneAll)
    {
        if (remote.IsNullOrEmpty())
        {
            // Fetch and prune AllRemotes
        }
        else
        {
            // Fetch and prune remote
        }

```

RussKie replied:
Apologies for being unclear.
I'd like to eliminate `FetchPrune` altogether, the implementation looks similar to `FetchPruneAll`.

mast-eu replied:
Indeed, they are very similar. I can make it work without `FetchPrune`.

mast-eu replied:
Done, please have a look.

RussKie commented:
@pmiossec are you happy with the changes?

pmiossec commented:
>@pmiossec are you happy with the changes?

Yes. The only thing that we could improve is that we could remove the last use of `FormRemoteProcess.ShowDialog(this, "remote update");` in 
https://github.com/gitextensions/gitextensions/blob/master/GitUI/CommandsDialogs/FormRemotes.cs#L450
to also use `UICommands.StartPullDialogAndPullImmediately()`.

mast-eu commented:
> The only thing that we could improve is that we could remove the last use of `FormRemoteProcess.ShowDialog(this, "remote update");` in
> https://github.com/gitextensions/gitextensions/blob/master/GitUI/CommandsDialogs/FormRemotes.cs#L450
> to also use `UICommands.StartPullDialogAndPullImmediately()`.

Good point. I will add this.

mast-eu commented:
I added a commit implementing @pmiossec's suggestion.

RussKie commented:
Thank you
